kotlinx-coroutines-core/jvm/test/LimitedParallelismStressTest.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2016-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines
+
+import org.junit.*
+import org.junit.Test
+import org.junit.runner.*
+import org.junit.runners.*
+import java.util.concurrent.atomic.*
+import kotlin.test.*
+
+@RunWith(Parameterized::class)
+class LimitedParallelismStressTest(private val targetParallelism: Int) : TestBase() {
+
+    companion object {
+        @Parameterized.Parameters(name = "{0}")
+        @JvmStatic
+        fun params(): Collection<Array<Any>> = listOf(1, 2, 3, 4).map { arrayOf(it) }
+    }
+
+    @get:Rule
+    val executor = ExecutorRule(targetParallelism * 2)
+    private val iterations = 100_000 * stressTestMultiplier
+
+    private val parallelism = AtomicInteger(0)
+
+    private fun checkParallelism() {
+        val value = parallelism.incrementAndGet()
+        assertTrue { value <= targetParallelism }
+        parallelism.decrementAndGet()
+    }
+
+    @Test
+    fun testLimitedExecutor() = runTest {
+        val view = executor.limitedParallelism(targetParallelism)
+        repeat(iterations) {
+            launch(view) {
+                checkParallelism()
+            }
+        }
+    }
```

Comment:
We could also check that the target parallelism is reached eventually. I see some similar checks for the default dispatcher, but maybe it's worth checking in the general case. It's probably not a stress-test though.
```kotlin
    @Test
    fun testLimitedExecutorReachesTargetParallelism() = runTest {
        val view = executor.limitedParallelism(targetParallelism)
        val latch = CountDownLatch(1)
        val parallelism = AtomicInteger(0)
        val jobs = (1..targetParallelism * 2).map {
            launch(view) {
                parallelism.incrementAndGet()
                latch.await()
            }
        }
        while (parallelism.get() < targetParallelism) {
            Thread.sleep(10)
        }
        Thread.sleep(10)
        assertTrue { parallelism.get() == targetParallelism }
        latch.countDown()
        jobs.joinAll()
    }
```

Reply:
Thanks for the suggestion, added it in the form of a stress test